gitlabform/processors/project/schedules_processor.py

```diff
@@ -42,102 +48,124 @@ def _process_configuration(self, project_and_group: str, configuration: dict):
-                        ),
+                    schedule = project.pipelineschedules.get(schedule_ids[0])
+                    self._update_existing_schedule(
+                        configured_schedules, project, schedule, schedule_description
                     )
                 elif schedule_ids:
                     debug(
                         "Replacing existing pipeline schedules '%s'",
                         schedule_description,
                     )
                     for schedule_id in schedule_ids:
-                        self.gitlab.delete_pipeline_schedule(
-                            project_and_group, schedule_id
-                        )
-                    self.create_schedule_with_variables(
-                        configuration, project_and_group, schedule_description
+                        project.pipelineschedules.get(schedule_id).delete()
+
+                    self._create_schedule_with_variables(
+                        configured_schedules, project, schedule_description
                     )
                 else:
                     debug("Creating pipeline schedule '%s'", schedule_description)
-                    self.create_schedule_with_variables(
-                        configuration, project_and_group, schedule_description
+                    self._create_schedule_with_variables(
+                        configured_schedules, project, schedule_description
                     )
 
         if enforce_schedules:
             debug("Delete unconfigured schedules because enforce is enabled")
 
-            for schedule in existing_schedules:
-                schedule_description = schedule["description"]
-                schedule_id = schedule["id"]
+            self._delete_schedules_no_longer_in_config(
+                configured_schedules, existing_schedules, project
+            )
+
+    def _update_existing_schedule(
+        self,
+        configured_schedules,
+        project: Project,
+        schedule: ProjectPipelineSchedule,
+        schedule_description: str,
+    ):
+        entity_config = configured_schedules[schedule_description]
+        if self._needs_update(entity_config, schedule.asdict()):
+            # Delete and then re-create schedule so we can pass all info in the data to Gitlab in case their APIs change
+            project.pipelineschedules.delete(schedule.id)
 
-                debug(f"processing {schedule_id}: {schedule_description}")
-                if schedule_description not in configured_schedules:
-                    debug(
-                        "Deleting pipeline schedule named '%s', because it is not in gitlabform configuration",
-                        schedule_description,
-                    )
-                    self.gitlab.delete_pipeline_schedule(project_and_group, schedule_id)
+            self._create_schedule_with_variables(
+                configured_schedules, project, schedule_description
+            )
```

Comment:
Why do we need to delete the existing schedule and create a new one? Can't we just perform an update? The API supports a `PUT` operation. That's what the current logic is uses.

https://github.com/gitlabform/gitlabform/blob/69cdd7ab49a0b1516ddba5be4ada5760dc742041/gitlabform/gitlab/schedules.py#L36-L45

Creating a new schedule might be problematic. Each schedule has a separate ID. Imagine running gitlabform periodically. There'll be a ton of records of schedule in the database because of constant delete and create new. Doing this will also make the following API useless to end users because they will lose the history.

https://docs.gitlab.com/ee/api/pipeline_schedules.html#get-all-pipelines-triggered-by-a-pipeline-schedule

So, I don't think updating schedule should require deleting the current one and creating a new one. Let me know if there's something I'm unaware of that requires doing this.

Reply:
hmm yeah good point, I will need to have a look and see if there's a way to pass all the data in the update request

Reply:
Hi @TimKnight-Opencast . Will you have opportunity to look into this? You're helping to refactor the group_member processor in another PR that's also in progress. Appreciate it btw.

I feel that one is a bit more sensitive/complex than this one and might take a bit of time to review. Maybe we can get this completed and then focus on the other one.

To me this one seems very close to be done. That's why I'm suggesting to finish it first. But if you prefer to come back to this later, that's fine too. Thanks again for everything.

Reply:
@amimas - on this now, going to update the tests to verify when updating a schedule the updated schedule has the same id as the existing schedule (i.e. wasn't deleted and then recreated)

Reply:
> going to update the tests to verify when updating a schedule the updated schedule has the same id as the existing schedule (i.e. wasn't deleted and then recreated)

Awesome! Thank you so much! Looking forward to the update.

Reply:
Let's close this. Thanks for the refactor and updating the tests.